components/script/dom/promisedebugging.rs

```diff
@@ -0,0 +1,52 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use dom::bindings::inheritance::Castable;
+use dom::bindings::root::DomRoot;
+use dom::event::{Event, EventBubbles, EventCancelable};
+use dom::eventtarget::EventTarget;
+use dom::globalscope::GlobalScope;
+use dom::promise::Promise;
+use dom::promiserejectionevent::PromiseRejectionEvent;
+use js::jsapi::{GetPromiseResult, HandleObject};
+use js::rust::HandleValue;
+use js::rust::wrappers::JS_ErrorFromException;
+use std::cell::Cell;
+
+pub struct PromiseDebugging {
+  pub id_predix: Cell<String>,
+}
+
+impl PromiseDebugging {
+    #[allow(unsafe_code)]
+    pub fn AddUncaughtRejection(global: DomRoot<GlobalScope>, promise: HandleObject) {
+        global.set_uncaught_rejections(promise);
+
+        unsafe {
+            PromiseDebugging::flush_uncaught_rejections_internal(global, promise);
+        }
+    }
+
+    #[allow(unsafe_code)]
+    pub unsafe fn flush_uncaught_rejections_internal(global: DomRoot<GlobalScope>, promise: HandleObject) {
+        let cx = global.get_cx();
+
+        let result = GetPromiseResult(promise);
+        rooted!(in(cx) let object = result.to_object());
+        let error_report = JS_ErrorFromException(cx, object.handle());
+        println!("{:?}, {:?}", result, error_report);
+
+        let event = PromiseRejectionEvent::new(
+            &global.upcast::<GlobalScope>(),
+            atom!("unhandledrejection"),
+            EventBubbles::DoesNotBubble,
+            EventCancelable::Cancelable,
+            // FIXME(CYBAI): Use real rejected `promise` and `reason`
+            Promise::new(&global.upcast::<GlobalScope>()),
+            HandleValue::null()
```

Comment:
@jdm , @nox :

I'm stuck by how to convert a `HandleObject` into a `Rc<Promise>` and also `Value` into `HandleValue` here.

Reply:
We can use `Promise::new_with_js_promise` here if we make the method public.

Reply:
Yes, `new_with_js_promise` is what I should use and just make it as a `pub`lic method.

Previously, I also tried this method but the `assert` inside `new_with_js_promise` complained about that the passed `HandleObject` is not `Promise`. But now it works fine! I think I used wrong `HandleObject` before 😓